sqlite3_go18_test.go

```diff
@@ -136,6 +136,89 @@ func TestShortTimeout(t *testing.T) {
 	}
 }
 
+func TestBeginTxCancel(t *testing.T) {
+	srcTempFilename := TempFilename(t)
+	defer os.Remove(srcTempFilename)
+
+	db, err := sql.Open("sqlite3", srcTempFilename)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	db.SetMaxOpenConns(10)
+	db.SetMaxIdleConns(5)
+
+	defer db.Close()
+	initDatabase(t, db, 100)
+
+	wg := sync.WaitGroup{}
+	// create several go-routines to expose racy issue
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for i := 0; i < 1000; i++ {
+				ctx, cancel := context.WithCancel(context.Background())
+				go cancel() // make it cancel concurrently with exec("BEGIN");
```

Comment:
I wonder this canceliing always works as you expected. Sleep is needed?

Reply:
Unfortunately, the code being tested here is inherently non-deterministic. @azavorotnii Did this test as written fail consistently prior to your changes?

Reply:
yes, it failed every time I tested it without fix in code, but I agree that "passed" status doesn't prove issue absence.